source/beamline/SGM/energy/SGMEnergyControlSet.cpp

```diff
@@ -0,0 +1,336 @@
+		onUndulatorPositionPVValueChanged(undulatorPosition()->value());
+		onUndulatorDetuneOffsetPVValueChanged(undulatorDetuneOffset()->value());
+		onUndulatorHarmonicPVValueChanged(undulatorHarmonic()->value());
+		onUndulatorTrackingPVValueChanged(undulatorTracking()->value());
+		onExitSlitPositionPVValueChanged(exitSlitPosition()->value());
+		onExitSlitTrackingPVValueChanged(exitSlitPositionTracking()->value());
+
+		emit initialized();
+	}
+}
+
+void SGMEnergyControlSet::onEnergyPVValueChanged(double value)
+{
+	energyPositionValidator_->requestEnergy(value);
+}
+
+void SGMEnergyControlSet::onGratingAnglePVValueChanged(double value)
+{
+	energyPositionValidator_->setGratingAngle(value);
+}
+
+void SGMEnergyControlSet::onGratingTranslationPVValueChanged(double)
+{
+	SGMGratingSupport::GratingTranslation newGratingTranslation;
+
+	if(gratingTranslation()->withinTolerance(0)) {
+
+		newGratingTranslation = SGMGratingSupport::LowGrating;
+	} else if (gratingTranslation()->withinTolerance(1)) {
+
+		newGratingTranslation = SGMGratingSupport::MediumGrating;
+	} else if(gratingTranslation()->withinTolerance(2)) {
+
+		newGratingTranslation = SGMGratingSupport::HighGrating;
+	} else {
+
+		newGratingTranslation = SGMGratingSupport::UnknownGrating;
+	}
+
+	energyPositionValidator_->setGratingTranslation(newGratingTranslation);
+}
+
+void SGMEnergyControlSet::onGratingtranslationOptimizationPVValueChanged(double)
+{
+	SGMEnergyPosition::GratingTranslationOptimizationMode newOptimizationMode;
+
+	if(gratingTranslationOptimization()->withinTolerance(0)) {
+
+		newOptimizationMode = SGMEnergyPosition::ManualMode;
+	} else if(gratingTranslationOptimization()->withinTolerance(1)) {
+
+		newOptimizationMode = SGMEnergyPosition::OptimizeFlux;
+	} else if(gratingTranslationOptimization()->withinTolerance(2)) {
+
+		newOptimizationMode = SGMEnergyPosition::OptimizeResolution;
+	} else {
+
+		return;
+	}
+
+	energyPositionValidator_->setGratingTranslationOptimizationMode(newOptimizationMode);
+}
+
+void SGMEnergyControlSet::onUndulatorPositionPVValueChanged(double value)
+{
+	energyPositionValidator_->setUndulatorPosition(value);
+}
+
+void SGMEnergyControlSet::onUndulatorDetuneOffsetPVValueChanged(double value)
+{
+	energyPositionValidator_->setUndulatorOffset(value);
+}
+
+void SGMEnergyControlSet::onUndulatorHarmonicPVValueChanged(double)
+{
+	SGMUndulatorSupport::UndulatorHarmonic newHarmonic;
+
+	if(undulatorHarmonic()->withinTolerance(0)) {
+
+		newHarmonic = SGMUndulatorSupport::FirstHarmonic;
+	} else if(undulatorHarmonic()->withinTolerance(1)) {
+
+		newHarmonic = SGMUndulatorSupport::ThirdHarmonic;
+	} else {
+
+		newHarmonic = SGMUndulatorSupport::UnknownUndulatorHarmonic;
+	}
+
+	energyPositionValidator_->setUndulatorHarmonic(newHarmonic);
+}
+
+void SGMEnergyControlSet::onUndulatorTrackingPVValueChanged(double)
+{
+	bool isTracking;
+
+	if(undulatorTracking()->withinTolerance(0)) {
+
+		isTracking = false;
```

Comment:
Technically, lines like this could be simplified to `isTracking = !undulatorTracking()->withinTolerance(0);`